src/contexts/ethereum-provider.context.tsx

```diff
@@ -0,0 +1,119 @@
+  const env = useEnvContext();
+
+  const connectProvider = async (walletName: WalletName): Promise<void> => {
+    switch (walletName) {
+      case WalletName.METAMASK: {
+        if (window.ethereum) {
+          const metaMaskProvider = window.ethereum;
+          const web3Provider = new Web3Provider(window.ethereum);
+
+          return getConnectedAccounts(web3Provider)
+            .then((accounts) => {
+              setRawProvider(metaMaskProvider);
+              setProvider(web3Provider);
+              setAccount({ status: "successful", data: accounts[0] });
+            })
+            .catch(() => setAccount({ status: "failed", error: "Error loading account" }));
+        } else {
+          return setAccount({
+            status: "failed",
+            error: "There is no Ethereum provider injected in the window",
+          });
+        }
+      }
+      case WalletName.WALLET_CONNECT: {
+        if (env) {
+          const walletConnectProvider = new WalletConnectProvider({
+            infuraId: env.REACT_APP_INFURA_API_KEY,
+          });
+          const web3Provider = new Web3Provider(walletConnectProvider);
+
+          return walletConnectProvider
+            .enable()
+            .then((accounts) => {
+              setRawProvider(walletConnectProvider);
+              setProvider(web3Provider);
+              setAccount({ status: "successful", data: accounts[0] });
+            })
+            .catch(() => {
+              setAccount({ status: "failed", error: "Error loading account" });
+            });
+        } else
+          return setAccount({
+            status: "failed",
+            error: "The env hasn't be initialized properly",
+          });
+      }
+    }
+  };
+
+  useEffect(() => {
+    const onAccountsChanged = (accounts: unknown): void => {
+      const parsedAccounts = ethereumAccountsParser.safeParse(accounts);
+
+      if (parsedAccounts.success) {
+        if (parsedAccounts.data.length) {
+          setAccount({ status: "successful", data: parsedAccounts.data[0] });
+        } else {
+          setAccount({ status: "pending" });
+          navigate(routes.login.path);
+        }
+      }
+    };
+    const onChainChangedOrDisconnect = () => {
+      window.location.reload();
+    };
+
+    if (rawProvider && rawProvider.on) {
+      rawProvider.on(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
+      rawProvider.on(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
+      rawProvider.on(EthereumEvent.DISCONNECT, onChainChangedOrDisconnect);
+    }
+
+    return () => {
+      if (rawProvider && rawProvider.removeListener) {
+        rawProvider.removeListener(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
+        rawProvider.removeListener(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
+        rawProvider.removeListener(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
```

Comment:
It looks like here we should be using the instance of the class `Web3Provider` instead of the `ExternalProvider` object we pass to the constructor.

I guess we've had to create the interface `MetaMaskProvider` just to be able to manage the events here. There's another comment regarding that interface.

Reply:
After our meeting, I understood the reasoning for the current implementation: `Web3Provider` instances do not support the [EIP-1193 events](https://eips.ethereum.org/EIPS/eip-1193#events) `on` and `removeListener`.

We could though reach the `ExternalProvider` present in the `Web3Provider` instance at `provider.provider` and make sure the methods we need exist:

```typescript
if (provider) {
  const rawProvider: Record<string, unknown> = provider.provider;
  if ("on" in rawProvider && typeof rawProvider.on === "function") {
    rawProvider.on(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
    rawProvider.on(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
    rawProvider.on(EthereumEvent.DISCONNECT, onChainChangedOrDisconnect);
  }
}

return () => {
  if (provider) {
    const rawProvider: Record<string, unknown> = provider.provider;
    if ("removeListener" in rawProvider && typeof rawProvider.removeListener === "function") {
      rawProvider.removeListener(EthereumEvent.ACCOUNTS_CHANGED, onAccountsChanged);
      rawProvider.removeListener(EthereumEvent.CHAIN_CHANGED, onChainChangedOrDisconnect);
      rawProvider.removeListener(EthereumEvent.DISCONNECT, onChainChangedOrDisconnect);
    }
  }
};
```

This approach is equally typesafe and has the advantage of limiting dirtiness to where we need to call the methods, in contrast to having to create the `MetaMaskProvider` interface and store the `rawProvider` in the context.

Reply:
Done!